[docs-infra] Add a collapsible list & table views to the API content display

alexfauquette opened the pull request:
## Preview

- [Material UI](https://deploy-preview-38265--material-ui.netlify.app/material-ui/api/autocomplete/#Autocomplete-css-popupIndicator)
- [Joy UI](https://deploy-preview-38265--material-ui.netlify.app/joy-ui/api/autocomplete/)
- [Base UI](https://deploy-preview-38265--material-ui.netlify.app/base-ui/react-menu/components-api/#menu)
- [MUI System](https://deploy-preview-38265--material-ui.netlify.app/system/api/container/)
- [Data grid premium](https://deploy-preview-38265--material-ui.netlify.app/experiments/docs/data-grid-premium/)

oliviertassinari commented:
I chatted a bit with @alexfauquette, I think that it could indeed be great to run an A/B on this. We could have 3 implementations:

1. Current
2. This PR
3. Blueprint/Stripe way https://blueprintjs.com/docs/#datetime/date-input.props-interface. For very long props, we could break them in two lines, we shouldn't have very long props anyway.

and see which one the developers pick the most as their default. Personally, as a developer, I would definitely put 2. last, too much effort to click, scrolling is easier. I would have to try 1 and 3 to really know for sure, maybe a slight preference on 3.

danilo-leal commented:
@oliviertassinari I recently wrote [a doc about the API content display](https://www.notion.so/mui-org/API-content-display-design-11697af821144e9387a621ee59221414) and the A/B test idea could maybe be discussed more in-depth there! I personally don't see a lot of value in testing something close to Blueprint but happy to further discuss on Notion considering what Alex & I discussed that's documented there! 😃

joserodolfofreitas commented:
> Blueprint/Stripe way https://blueprintjs.com/docs/#datetime/date-input.props-interface. For very long props, we could break them in two lines, we shouldn't have very long props anyway.

I like the UX on this one, it solves the main problem with the current design (which mixes secondary with primary information on the same vertical space and breaks quick skimming). Overall I think the card design is good, I'd bet we just need to keep the primary information quickly scannable.
Also to consider: We could collapse options by default only if they cross the max size threshold.

alexfauquette commented:
To simplify experimentation, I added a page with data grid premium API docs such that we can see in live the impact on big API page

https://deploy-preview-38265--material-ui.netlify.app/experiments/docs/data-grid-premium/

Based on the comments "I want to easily scan the props or description", I moved to a table layout with props/description column

I allow a first column of up to 200px. If the props overflow, their name is cut in two (mainly for data grid unstable props)


Props now have a check box on top of it to enable/disable collapsed rows


## Todo

- [ ] Adapt other sections to collapse
- [ ] Setup the localhost for saving preference
- [ ] Improve the UI for selection between collapsable or not
- [ ] Setup Google Analytics to track UI preferences
- [ ] Bring back the table rendering

oliviertassinari commented:
@alexfauquette The plan sounds great.

I think this is a viable implementation to test 3. in https://github.com/mui/material-ui/pull/38265#issuecomment-1664235088. I have added a screenshot to illustrate it. I would only raise that the prop description font is wrong, it uses `theme.typography.fontFamilyCode` rather than `theme.typography.fontFamily`.

alexfauquette commented:
## Open to discussion

### Settings granularity

For now, I plan to allow to toggle between collapse/expended on a section type level. This means you could be fully expended on CSS and collapsed in props. This should allow us to get finer statistics about users' preferences.

### Layout

The green information ("State" for CSS and "Require" for props) could be moved below the title

![image](https://github.com/mui/material-ui/assets/45398769/b3649ee9-187e-4e04-bfbf-d0c4f9d08c43)

### What to hide

I went with a simple solution: only show the description when collapsed. But it can be adapted: To hide something, just add className `MuiApi-collapsible` to the element

## To do

- [ ] Update analytics according to the granularity chosen
- [ ] Clean CSS
	- [ ] Fix font
	- [ ] Fis margins when collapsed/extended
- [ ] Clean the ToC
    - [ ] Allows collapsing for props and CSS
    - [ ] Avoid overflow (especially for data grid)

danilo-leal commented:
I've meant to get back to this for some time now and finally managed to scoop some time for it last week! Overall, I'm a bit unsure about this design⎯will try to break it down:

1. With the two columns, I think we start to get back to the table layout without fully committing to it. In my perspective, that's making cases where the list layout is better (e.g., X components) worse and cases where the table layout might be better (every other simpler component, as per much of the feedback received so far 😅) worse, because this version of a table is not as polished.
2. Not confident about the dense vs. not dense (i.e., accordion vs. not accordion) toggle. I think I'd rather go all in and do a toggle between list and table, as we [**discussed in the Notion doc**](https://www.notion.so/mui-org/API-content-display-design-11697af821144e9387a621ee59221414?pvs=4#eb33b9a4a4ea4caa814c2e5c156a3184). I acknowledge it might be more complex but I think it'd make for a better design/experience as a whole.

I played around with [a few iterations on Figma](https://www.figma.com/file/MfirV6NOFAp2xxAZX1QU1V/Documentation?type=design&node-id=2020%3A403&mode=design&t=NOKcsBgC9SqkI7I1-1), still using the list layout but experimenting with an accordion. My goal was to find the most scannable option, as it's what I've understood to be the most important thing as per the feedback, which is why I enjoyed the below-attached option the most, even though there are trade-offs. 

<img width="400" alt="Screen Shot 2023-09-11 at 18 42 12" src="https://github.com/mui/material-ui/assets/67129314/a43eabe2-1b15-4e41-8aa3-3372c5110958">


Even put [a message up on the Discord docs-infra channel](https://discord.com/channels/1131323012554174485/1150869423398330438/1150869592978243664) to get some feedback about it... it might work, but, at the end of the day, I still think the table vs. list toggle would be the most effective solution, enabling folks to choose their preferred view while knowing that each one of them is better in different scenarios. Let me know what you think about it!

alexfauquette commented:
> 3. A potential reason as to why I don't feel confident would be that in the dense (accordion) view, if the description is what persists in the non-dense view, we'd have cases (like the AccordionSummary attached below) where it doesn't get dense at all; very little visible change, at a glance. If we were to move forward with this, I'd maybe inverse things out and have the description hidden in the accordion rather than the default value + type info.

I fully agree with this observation. Recently I thought about displaying everything, and use `max-height` plus `overflow: hidden` to set a dense row.

Something in that taste, but of course with some gradient and better spacing 🙈

![image](https://github.com/mui/material-ui/assets/45398769/53fbd5c0-2486-49f1-b504-7f7e05ed64dd)


About the two-columns aspect, the reason was mostly to allow a preview. If you're unsure about what the props are doing sometimes just the beginning of the description is enough. Then you can expend it, or continue searching a the prop you need.

## Moving forward

What we learned is that API page are mostly to discover props. If you want the definition, it's already available in the IDE. The only case where you can note use your IDE is when you don't remember the props name and so don't know what to type.

The list design without collapse does not solve the issue for data grid since it implies scrolling a lot. The navigation menu solves this issue in a hacky way.

Proposal: We could move to the iteration #6. For me, it's fine to only have the name. For bringing back table, I'm not sure about how to proceed. Do you want to first work on a design in figma, or I can setup the infra for the table display, and let you do the CSS to see directly how does it render

danilo-leal commented:
> Proposal: We could move to the iteration 6. For me, it's fine to only have the name. For bringing back table, I'm not sure about how to proceed. Do you want to first work on a design in figma, or I can setup the infra for the table display, and let you do the CSS to see directly how does it render

@alexfauquette I'm happy either way, honestly! I believe the table <> list toggle route to be the most effective in the long term, leaving me wondering if we shouldn't step right in with it. But... that could be a bit more complex while experimenting with iteration 6 seems _much_ faster. So, what if we did the latter and used the deploy preview link to gather feedback before merging/going all in?

alexfauquette commented:
I've removed the 2 column design for something closer to iteration number 6.

I've also added the support for table display.

The new architecture is the following:

For each section (props, CSS, slots, ...), there is one component that manages the rendering (`<PropertiesSection />` for props)

This component is responsible to
- Display helper texts
- Chose between listed and table display
- Aggregate all the data about the API section and format them in a format easy to display (`formatedProperties`)

Then there are two distinct components `<PropertiesTable/>` and `<PropertiesList/>` which take this `formatedProperties` and do the rendering

I only did the props (but it is the most problematic one). It's working. I will try to give it a bit more love during the next week (clean the CSS since it suffers the many different experiments, make sure it works on all API pages, support also other sections)

alexfauquette commented:
@danilo-leal Thanks for the table styling, it already looks nice ✨

 I finished the new file structure (classes, slots, and CSS are now available in list and table format). Now the folder structure is the following:

```
- ApiPage
  - list
  - table
  - sections
```

For the moment I just copy pasted the table styling. Might need to think about what is shared and is not between different sections (but it's the end of the day for me 😁)

danilo-leal reviewed (approved):
I'm really happy with how this one turned out! 😍 I'd be happy to merge it the way it is and have any other addition be an iterative improvement — @alexfauquette I _really_ appreciate all the work and consideration for the ideas; awesome work! 🙌

alexfauquette commented:
🎉🎉 Celebration time 🎉🎉

I just need to review my own code to make sure the data are correctly sent to Google Analytics, such that we can do something with them. Should be ready for the next release 🚀
